File: java-simple-http-server/src/main/java/simplehttpserver/Response.java
package simplehttpserver;

import simplehttpserver.enums.Status;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

public class Response {
    final Status status;
    final String contentType;
    final byte[] body;

    private static final DateTimeFormatter rfc1123Formatter = DateTimeFormatter.RFC_1123_DATE_TIME;
    private static final String CRLF = "\r\n";

    public Response(Status status, String contentType, byte[] body) {
        this.status = status;
        this.contentType = contentType;
        this.body = body;
    }

    public byte[] toBytes() {
        OffsetDateTime now = OffsetDateTime.now(ZoneOffset.UTC);
        var header =  "HTTP/1.1 " + status.statusCode + CRLF +
                "Date: " + rfc1123Formatter.format(now) + CRLF +
                "Server: SimpleJavaHttpServer" + CRLF +
                "Content-Type: " + contentType + CRLF +
                "Content-Length: " + body.length + CRLF +
                "Connection: Close" + CRLF +
                CRLF;
        var headerBytes = header.getBytes(StandardCharsets.UTF_8);
        var buff = ByteBuffer.allocate(headerBytes.length + body.length);
        buff.put(headerBytes);
        buff.put(body);
        return buff.array();
    }
}

File: scala-simple-http-server/src/main/scala/simplehttpserver/Response.scala
package simplehttpserver

import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.time.format.DateTimeFormatter
import java.time.{OffsetDateTime, ZoneOffset}

case class Response(status: Status, contentType: String, body: Array[Byte]) {

  import Response._

  def toBytes: Array[Byte] = {
    val now = OffsetDateTime.now(ZoneOffset.UTC)
    val response =
      s"""HTTP/1.1 ${status.value}
         |Date: ${rfc1123Formatter.format(now)}
         |Server: SimpleScalaHttpServer
         |Content-Type: $contentType
         |Content-Length: ${body.length.toString}
         |Connection: Close
         |
         |""".stripMargin

    val headerBytes = response.getBytes(StandardCharsets.UTF_8)
    val buff = ByteBuffer.allocate(headerBytes.length + body.length)
    buff.put(headerBytes)
    buff.put(body)
    buff.array()
  }
}

object Response {
  private val HtmlMime = "text/html;charset=utf8"

  val BadRequest: Response =
    Response(
      status = Status.BadRequest,
      contentType = HtmlMime,
      body = "Bad Request".getBytes(StandardCharsets.UTF_8)
    )

  val Forbidden: Response =
    Response(
      status = Status.Forbidden,
      contentType = HtmlMime,
      body = "Forbidden".getBytes(StandardCharsets.UTF_8)
    )

  val NotFound: Response =
    Response(
      status = Status.NotFound,
      contentType = HtmlMime,
      body = "Not Found".getBytes(StandardCharsets.UTF_8)
    )

  private val rfc1123Formatter = DateTimeFormatter.RFC_1123_DATE_TIME
}
